cohortextractor2/dsl.py

```diff
@@ -161,9 +169,109 @@ class PatientSeries:
+        return self.value < self._get_other(other)
+
+    def __le__(self, other: Expression) -> Comparator:
+        return self.value <= self._get_other(other)
+
+    def __eq__(self, other: Expression) -> Comparator:  # type: ignore[override]
+        return self.value == self._get_other(other)
+
+    def __ne__(self, other: Expression) -> Comparator:  # type: ignore[override]
+        return self.value != self._get_other(other)
+
+    def __and__(self, other: Expression) -> Comparator:
+        return self.value & self._get_other(other)
+
+    def __or__(self, other: Expression) -> Comparator:
+        return self.value | self._get_other(other)
+
+    def __invert__(self) -> Comparator:
+        return ~self.value
+
+    def __repr__(self) -> str:
+        return f"PatientSeries(value={self.value})"
+
+
+Expression = Union[PatientSeries, Comparator]
+
+
+def categorise(mapping: dict[str, Expression], default: bool | None) -> PatientSeries:
+    """
+    Represents a switch statement.
+
+    mapping: A dict mapping category keys to category expressions, e.g.
+        {
+            "0-17": age < 18,
+            "18-24": (age >= 18) & (age < 25),
+            "25-34": (age >= 25) & (age < 35),
+            "35-44": (age >= 35) & (age < 45),
+            "45-54": (age >= 45) & (age < 55),
+            "55-69": (age >= 55) & (age < 70),
+            "70-79": (age >= 70) & (age < 80),
+            "80+": age >= 80,
+        }
+        where `age` is a PatientSeries
+        Category expressions can be Comparators generated by comparing a PatientSeries
+        with a constant, as above, or simple PatientSeries (which will be converted to a
+        not-null comparison)
+        {
+            "has_code": code,
+        }
+        where `code` is a PatientSeries, selecting the column "code"
+
+    returns: PatientSeries with a ValueFromCategory value
+    """
+    _validate_category_mapping(mapping)
+    mapping = {
+        key: not_null_comparator(mapped_value.value)
+        if isinstance(mapped_value, PatientSeries)
+        else mapped_value
+        for key, mapped_value in mapping.items()
+    }
+    return PatientSeries(value=ValueFromCategory(mapping, default))
+
+
+def _validate_category_mapping(mapping: dict[str, Expression]) -> None:
+    """
+    Ensure that a category mapping is valid, by checking that:
+    - there are no duplicate values
+    - all keys are the same
+    - all values are either PatientSeries or Comparator
+    """
+    seen_values = set()
+    duplicates = set()
+    for key, value in mapping.items():
+        if not isinstance(value, (Comparator, PatientSeries)):
+            raise TypeError(
+                "Category values must be either a PatientSeries, or a "
+                "comparison expression involving a PatientSeries. "
+                f"Got '{value}' ({type(value)}) for category key '{key}'"
+            )
+        if repr(value) in seen_values:
```

Comment:
Oh, ouch. I see why you've had to do this. Could we may be add a method for testing equality instead?